octave new version 10.2.0 demands stdc++17

cessenat opened the pull request:
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->
There are many changes:
 - removed maintainer @mtmiller that unfortunately is no longer active,
 - added maintainer @mmuetzel and myself
 - added version 10.2.0 that demands stdc++17 with GNU extensions,
 - OCTAVE_CONF_MKOCTFILE_CXX is set to `quote(self.compiler.cxx + " -std=gnu++17")` when 10.2.0
 - run environment sets CXX to `self.compiler.cxx + " -std=gnu++17"` when 10.2.0

I would like to have advise from many skilled spack packagers, such as @alalazo, @bernhardkaindl, @michaelkuhn, @alecbcs, @davydden since the way I made Octave 10.2.0 compile properly, with extensions using spack packages or using Octave "pkg install -forge extension" where extension may be "io", "gsl", "aduino", ..., well, this way may be naive or irrelevant.
I have been greatly helped by Markus Mützel @mmuetzel from Savannah who indicated that within Octave package installation for 10.2.0 may require, on some platforms, the setting:
```
setenv('CXX', '/usr/bin/clang++ -std=gnu++17');
pkg install -forge io

```
This is why I added a `setup_run_environment` function to set CXX value.
I have tested these modifications with gcc 10.5.0 and octave 10.2.0 or 9.4.0 on Ubuntu 20.04 with no external dep apart from compiler on
$ spack debug report
* **Spack:** 1.1.0.dev0 (https://github.com/spack/spack/commit/4a3df09f488e4475692b7512df64318e068fe4ab)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/07dbc55cea8ae969762a8824fbe62b77d0adb6d3
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-zen2


This PR is related to discussion/issue
https://github.com/spack/spack-packages/issues/1451

Octave is such a thing, and so is spack, it is really worth.

cessenat committed:
octave new version 10.2.0 demands stdc++17

cessenat committed:
octave runtime for pkg install demands stdc++17 when 10.2.0 only

cessenat committed:
spurious new line removed

alecbcs commented:
I believe the "correct" way to do this now that compilers are nodes that implement language dependencies is
```python3
depends_on("cxx@17:", when="@10:")
```

I'd try that as an alternative to your modifications here and let me know if it doesn't work.

cessenat committed:
explicit cxx 17 dependency when 10.2.0 and above

cessenat commented:
> ```python
> depends_on("cxx@17:", when="@10:")
> ```

Good thing to explicitly set the dependency to C++ 17. Unluckily it happened to be unsufficient when I removed the other explicit `-std=gnu++17` modifications.

alalazo commented on repos/spack_repo/builtin/packages/octave/package.py:
```diff
@@ -86,6 +87,7 @@ class Octave(AutotoolsPackage, GNUMirrorPackage):
     # Required dependencies
     depends_on("c", type="build")
     depends_on("cxx", type="build")
+    depends_on("cxx@17:", when="@10:")
```
Let's not add this, since this is bound to fail when other packages will require e.g. `cxx@14` and are mixed in the same DAG. We discussed sometimes ago that the best way to deal with this is to extend virtual packages with variants, but this is not implemented yet.

cessenat replied:
OK, removed.

alalazo replied:
Thanks!

cessenat committed:
remove explicit cxx 17 dependency when 10.2.0 and above for later variant implementation

mmuetzel commented:
I've never used spack. So, I don't think I can act as a maintainer for this package. 
Feel free to tag me though if there are questions regarding Octave itself.

mmuetzel commented on repos/spack_repo/builtin/packages/octave/package.py:
```diff
@@ -132,9 +133,13 @@ def patch(self):
         mkoctfile_in = os.path.join(self.stage.source_path, "src", "mkoctfile.in.cc")
         quote = lambda s: '"' + s + '"'
+        if self.spec.satisfies("@10.2.0:"):
+            cxx = quote(self.compiler.cxx + " -std=gnu++17")
+        else:
+            cxx = quote(self.compiler.cxx)
         entries_to_patch = {
             r"%OCTAVE_CONF_MKOCTFILE_CC%": quote(self.compiler.cc),
-            r"%OCTAVE_CONF_MKOCTFILE_CXX%": quote(self.compiler.cxx),
+            r"%OCTAVE_CONF_MKOCTFILE_CXX%": cxx,
```
I'm not sure this is the best solution in general.
During `configure` Octave checks if the used compiler defaults to a new enough version of C++. If it doesn't, it adds flags that enable support for C++17 (preferably with GNU extensions).
If the compiler defaults to a newer version of C++ (e.g., C++23), it doesn't add any flags.

Whichever flags were necessary are stored with the compiler in `OCTAVE_CONF_MKOCTFILE_CXX` and used when a user builds .oct file binaries with `mkoctfile`.
These .oct file binaries should be built using the same compiler and C++ version as Octave itself.

Using the current approach, Octave could have been built using C++23 (using a new enough compiler), but .oct file sources would be built using C++17. That probably shouldn't be the default.

I still don't understand why the `mkoctfile.in.cc` file needs to be patched like that in the first place. What would happen if you didn't patch it at all?

cessenat replied:
Thus is done from
https://github.com/spack/spack-packages/commit/b0ae12637e6a423cebc4c98b173916aeb2b397a7
commit by @alalazo.
Setting " -std=gnu++17" is a workaround to be able to compile packages. For Octave it is not necessary.
Also, @alalazo explained "We discussed sometimes ago that the best way to deal with this is to extend virtual packages with variants, but this is not implemented yet.", so you are right this is probably not the best solution to keep. So either we wait for a smart solution and stick down to 9.4.3, or we do that ugly stuff for the moment to allow 10.2.0. What do spack experts think about that ?

mmuetzel replied:
I've read "compiler wrappers" being mentioned repeatedly. But what are these compiler wrappers? And why is it an issue if they are used?

In any case, this is only an issue because your Clang is so old (version 10) and doesn't default to C++17 or newer (like LLVM version 16 or newer).
Are all spack packages built with such an old version of Clang?

mmuetzel replied:
Alternative idea (assuming these compiler wrappers cannot be used when building .oct files while running Octave for some reason):
Instead of patching `mkoctfile.in.cc`, run the `configure` script and use `sed` to replace the compiler wrappers in the generated `Makefile` (in the build tree) with the "actual" compilers.
E.g., something like this:
```
sed -i "s/MKOCTFILE_CC= ${cc_wrapper}/MKOCTFILE_CC= ${actual_cc}/g" Makefile
sed -i "s/MKOCTFILE_CXX = ${cxx_wrapper}/MKOCTFILE_CXX = ${actual_cxx}/g" Makefile
sed -i "s/MKOCTFILE_F77 = ${f77_wrapper}/MKOCTFILE_F77 = ${actual_f77}/g" Makefile
```
That should be retaining potential compiler flags in these Makefile variables.
After that, proceed building the project with `make`.

I don't know how to integrate that into this build script though. I also don't know how to query the paths of the compiler wrappers. (The latter are potentially passed to the `configure` script as command line arguments or they are set as environment variables when the `configure` script is being run.)

alalazo replied:
> I've read "compiler wrappers" being mentioned repeatedly. But what are these compiler wrappers? And why is it an issue if they are used?

https://spack.readthedocs.io/en/latest/packaging_guide_build.html#compiler-wrappers-and-flags

mmuetzel replied:
Thank you for the link.

I'm still not sure if I understand what these compiler wrappers are. Are they only "valid" while packaging and are no longer "available" after a user installs a spack package? Or is there something more to it why `mkoctfile` can't use them?

Could the Octave package have an explicit dependency on the compiler wrapper package, so the wrappers are installed when Octave is installed?

mmuetzel commented on repos/spack_repo/builtin/packages/octave/package.py:
```diff
@@ -389,3 +394,9 @@ def setup_dependent_package(self, module, dependent_spec):
         module.octave = Executable(join_path(self.spec.prefix.bin, "octave"))
+
+    def setup_run_environment(self, env):
+        if self.spec.satisfies("@10.2.0:"):
+            env.set("CXX", self.compiler.cxx + " -std=gnu++17")
```
Not sure if this is a good default in general. See above.

cessenat replied:
> Not sure if this is a good default in general. See above.

Well, without that pkg install -forge package fails.
Octave compilation is fine without setting -std=gnu++17, but the extended packages fails without that.

mmuetzel replied:
Like I wrote above, that is because of the old Clang that you are using.

cessenat committed:
Remove a maintainer

Following
https://github.com/spack/spack-packages/pull/1514#issuecomment-3314751368

cessenat commented:
This PR is not good enough, so it is closed.
